File: libraries/typescript/src/validate.ts
import Ajv2020, { type ValidateFunction } from "ajv/dist/2020.js";

import { envelopeSchema, streamSchema } from "./registry.js";
import { EnvelopeError, StreamError, SCHEMA_VERSION } from "./types.js";

const TS_RE = /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\.\d{3}Z$/;

// Schemas declare draft 2020-12 — use the matching Ajv build.
const ajv = new Ajv2020({ allErrors: true, strict: false });

let _envelopeValidator: ValidateFunction | undefined;
const _streamValidators: Map<string, ValidateFunction> = new Map();

function envelopeValidator(): ValidateFunction {
  if (!_envelopeValidator) {
    _envelopeValidator = ajv.compile(envelopeSchema());
  }
  return _envelopeValidator;
}

function streamValidator(stream: string): ValidateFunction {
  const cached = _streamValidators.get(stream);
  if (cached) return cached;
  const v = ajv.compile(streamSchema(stream));
  _streamValidators.set(stream, v);
  return v;
}

export function validateEnvelope(obj: unknown): asserts obj is Record<string, unknown> {
  if (typeof obj !== "object" || obj === null || Array.isArray(obj)) {
    throw new EnvelopeError("event must be a JSON object");
  }
  const validate = envelopeValidator();
  if (!validate(obj)) {
    const msg = (validate.errors || [])
      .slice(0, 5)
      .map((e) => `${e.instancePath || "/"}: ${e.message}`)
      .join("; ");
    throw new EnvelopeError(`envelope validation failed: ${msg}`);
  }
  const ev = obj as Record<string, unknown>;
  if (typeof ev.ts !== "string" || !TS_RE.test(ev.ts)) {
    throw new EnvelopeError(
      `ts must match YYYY-MM-DDTHH:MM:SS.sssZ, got ${JSON.stringify(ev.ts)}`,
    );
  }
  if (ev.schema_version !== SCHEMA_VERSION) {
    throw new EnvelopeError(
      `schema_version must be "${SCHEMA_VERSION}", got ${JSON.stringify(ev.schema_version)}`,
    );
  }
}

export function validateStreamPayload(stream: string, payload: unknown): void {
  const validate = streamValidator(stream);
  if (!validate(payload)) {
    const msg = (validate.errors || [])
      .slice(0, 5)
      .map((e) => `${e.instancePath || "/"}: ${e.message}`)
      .join("; ");
    throw new StreamError(`stream "${stream}" payload invalid: ${msg}`);
  }
}

export function validateEvent(obj: unknown): asserts obj is Record<string, unknown> {
  validateEnvelope(obj);
  const ev = obj as Record<string, unknown>;
  validateStreamPayload(ev.stream as string, ev.payload);
}

File: libraries/python/events_ndjson/_validate.py
"""Internal validation helpers shared by Writer and Reader."""

from __future__ import annotations

import re
from typing import Any, Dict

from jsonschema import Draft202012Validator

from events_ndjson.registry import envelope_schema, stream_schema
from events_ndjson.types import EnvelopeError, StreamError

_TS_RE = re.compile(r"^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\.\d{3}Z$")
_SCHEMA_VERSION = "events-ndjson/v1"

_envelope_validator = Draft202012Validator(envelope_schema())


def validate_envelope(obj: Any) -> None:
    """Raise EnvelopeError if obj is not a valid envelope."""
    if not isinstance(obj, dict):
        raise EnvelopeError("event must be a JSON object")
    errors = sorted(_envelope_validator.iter_errors(obj), key=lambda e: e.path)
    if errors:
        msgs = "; ".join(f"{list(e.path)}: {e.message}" for e in errors[:5])
        raise EnvelopeError(f"envelope validation failed: {msgs}")
    if not _TS_RE.match(obj["ts"]):
        raise EnvelopeError(f"ts must match YYYY-MM-DDTHH:MM:SS.sssZ, got {obj['ts']!r}")
    if obj["schema_version"] != _SCHEMA_VERSION:
        raise EnvelopeError(
            f"schema_version must be {_SCHEMA_VERSION!r}, got {obj['schema_version']!r}"
        )


def validate_stream_payload(stream: str, payload: Dict[str, Any]) -> None:
    """Raise StreamError if payload fails the stream schema."""
    schema = stream_schema(stream)
    validator = Draft202012Validator(schema)
    errors = sorted(validator.iter_errors(payload), key=lambda e: e.path)
    if errors:
        msgs = "; ".join(f"{list(e.path)}: {e.message}" for e in errors[:5])
        raise StreamError(f"stream {stream!r} payload invalid: {msgs}")


def validate_event(obj: Any) -> None:
    validate_envelope(obj)
    validate_stream_payload(obj["stream"], obj["payload"])
